File: Python/minSteps.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: minSteps.py
@Function: 只有两个键的键盘 动态规划
@Link: https://leetcode-cn.com/problems/2-keys-keyboard/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-16
"""


class Solution(object):
    def minSteps(self, n):
        ans = 0
        d = 2
        while n > 1:
            while n % d == 0:
                ans += d
                n /= d
            d += 1
        return ans


# --------------------------
if __name__ == "__main__":
    num = 3

    solution = Solution()
    min_steps = solution.minSteps(num)
    print(f"The solution of this problem is {min_steps}")

File: C++/minSteps.cpp
/**
 * @File    : minSteps.cpp
 * @Brief   : 只有两个键的键盘 动态规划
 * @Link    : https://leetcode-cn.com/problems/2-keys-keyboard/
 * @Author  : Wei Li
 * @Date    : 2021-07-16
*/

#include <iostream>
#include <vector>
#include <cmath>

class Solution
{
public:
    int minSteps(int n)
    {
        std::vector<int> dp(n + 1);
        int h = std::sqrt(n);
        for (int i = 2; i <= n; ++i)
        {
            dp[i] = i;
            for (int j = 2; j <= h; ++j)
            {
                if (i % j == 0)
                {
                    dp[i] = dp[j] + dp[i / j];
                    break;
                }
            }
        }
        return dp[n];
    }
};

// -------------------------------
int main(int argc, char **argv)
{
    int num = 3;

    Solution solution;
    auto min_steps = solution.minSteps(num);
    std::cout << "The solution of this problem is " << min_steps << std::endl;

    return 0;
}
